tests/integration/construct_component_methods/testcomponent/index.ts

```diff
@@ -0,0 +1,92 @@
+// Copyright 2016-2021, Pulumi Corporation.  All rights reserved.
+
+import * as pulumi from "@pulumi/pulumi";
+import * as provider from "@pulumi/pulumi/provider";
+
+interface ComponentArgs {
+    first: pulumi.Input<string>;
+    second: pulumi.Input<string>;
+}
+
+class Component extends pulumi.ComponentResource {
+    public readonly first!: pulumi.Output<string>;
+    public readonly second!: pulumi.Output<string>;
+
+    constructor(name: string, args: ComponentArgs, opts?: pulumi.ComponentResourceOptions) {
+        if (opts?.urn) {
+            args = <any>{
+                first: undefined,
+                second: undefined,
+            };
+        }
```

Comment:
This and the `if` on line 26 are annoying. This is to be able to support recreating an instance of `Component` that comes back as `__self__` resource reference. I wish there was a cleaner way to enable this for components, so authors won't have to handle this.

Reply:
Hm, yeah, these are bothersome. I'll give this some thought.

Reply:
Again, I think this is something that can be improved by importing the generated SDK for a component. While the component instance will be different from the provider component class, both should fulfill the same public interface (which is a requirement for methods I believe). Then we can add some framework magic to do something like `ComponentFunction.prototype.apply( this=SDKInstance, args...)`